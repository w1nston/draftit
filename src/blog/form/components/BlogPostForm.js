import React, { Component } from 'react';
import ReactQuill from 'react-quill';
import { css } from 'emotion';
import 'react-quill/dist/quill.snow.css';

const formStyle = css`
  display: flex;
  flex-direction: column;
`;

const formRowStyle = css`
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  margin-bottom: 16px;
  width: 80%;
`;

const inputStyle = css`
  border-radius: 4px;
  height: 36px;
  padding: 0 8px;
  width: 30%;
`;

const createButtonStyle = css`
  background-color: #3146aa;
  color: #fff;
  cursor: pointer;
  height: 40px;
  width: 20%;
`;

const lastRowStyle = css`
  display: flex;
  justify-content: flex-end;
  margin: 0 auto;
  width: 80%;
`;

const errorAlertStyle = css`
  font-weight: 700;
  color: #721c24;
  background-color: #f8d7da;
  border-color: #f5c6cb;
  text-align: center;
  padding: 0.75rem 1.25rem;
  border: 1px solid transparent;
  border-radius: 0.25rem;
`;

class BlogPostForm extends Component {
  state = {
    text: '', // TODO: get state from persistance layer
    title: '',
    error: null,
  };

  handleChange = text => {
    this.setState({ text });
  };

  handleTitleChange = event => {
    this.setState({ title: event.target.value });
  };

  handleCreate = event => {
    event.preventDefault();

    const { title, text } = this.state;

    if (title === '') {
      this.setState({ error: 'Title can not be empty!' });
      return;
    }

    if (text === '') {
      this.setState({ error: 'What is the point of an empty blog post!?' });
      return;
    }

    if (this.props.onSubmit) {
      this.props.onSubmit(this.state);
    }
    this.setState({ title: '', text: '' });
  };

  // TODO: Handle autosave
  render() {
    return (
      <form onSubmit={this.handleCreate} className={formStyle}>
        {this.state.error && (
          <p className={errorAlertStyle}>{this.state.error}</p>
        )}
        <h1>Insert thoughts here...</h1>
        <div className={formRowStyle}>
          <label htmlFor="title">Title</label>
          <input
            className={inputStyle}
            value={this.state.title}
            onChange={this.handleTitleChange}
            type="text"
            name="title"
          />
        </div>
        <div className={formRowStyle}>
          <label htmlFor="content">Content</label>
          <ReactQuill
            name="content"
            value={this.state.text}
            onChange={this.handleChange}
            placeholder="Start typing here..."
            modules={{
              toolbar: [
                ['bold', 'italic', 'underline', 'strike'], // toggled buttons
                ['blockquote', 'code-block'],

                [{ header: 1 }, { header: 2 }], // custom button values
                [{ list: 'ordered' }, { list: 'bullet' }],
                [{ script: 'sub' }, { script: 'super' }], // superscript/subscript
                [{ indent: '-1' }, { indent: '+1' }], // outdent/indent
                [{ direction: 'rtl' }], // text direction

                [{ size: ['small', false, 'large', 'huge'] }], // custom dropdown
                [{ header: [1, 2, 3, 4, 5, 6, false] }],

                [{ color: [] }, { background: [] }], // dropdown with defaults from theme
                [{ font: [] }],
                [{ align: [] }],

                ['clean'],
              ],
            }}
          />
        </div>
        <div className={lastRowStyle}>
          <button className={createButtonStyle} type="submit">
            Create
          </button>
        </div>
      </form>
    );
  }
}

export default BlogPostForm;
