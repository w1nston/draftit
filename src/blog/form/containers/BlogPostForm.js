import React from 'react';
import { withRouter } from 'react-router-dom';
import { Subscribe } from 'unstated';
import BlogPostForm from '../components/BlogPostForm';
import { PostsContainer } from '../../../store/blog/PostsContainer';

export const blogPostFormSubscription = () => (
  <BlogPostForm />
);

export const BlogPostFormSubscriber = props => (
  <Subscribe to={[PostsContainer]}>
    {blogPostFormSubscription}  
  </Subscribe>
);

export default withRouter(BlogPostFormSubscriber);
