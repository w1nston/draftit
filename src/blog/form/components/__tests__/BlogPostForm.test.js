import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import BlogPostForm from '../BlogPostForm';

describe('BlogPostForm', () => {
  it('renders', () => {
    const component = shallow(<BlogPostForm />);
    expect(toJson(component)).toMatchSnapshot();
  });
});
