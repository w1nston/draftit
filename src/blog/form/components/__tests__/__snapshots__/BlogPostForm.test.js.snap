// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogPostForm renders 1`] = `
<form
  className="css-1iyoj2o"
  onSubmit={[Function]}
>
  <h1>
    Insert thoughts here...
  </h1>
  <div
    className="css-kpu76b"
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <input
      className="css-38trz7"
      name="title"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="css-kpu76b"
  >
    <label
      htmlFor="content"
    >
      Content
    </label>
    <Quill
      modules={
        Object {
          "toolbar": Array [
            Array [
              "bold",
              "italic",
              "underline",
              "strike",
            ],
            Array [
              "blockquote",
              "code-block",
            ],
            Array [
              Object {
                "header": 1,
              },
              Object {
                "header": 2,
              },
            ],
            Array [
              Object {
                "list": "ordered",
              },
              Object {
                "list": "bullet",
              },
            ],
            Array [
              Object {
                "script": "sub",
              },
              Object {
                "script": "super",
              },
            ],
            Array [
              Object {
                "indent": "-1",
              },
              Object {
                "indent": "+1",
              },
            ],
            Array [
              Object {
                "direction": "rtl",
              },
            ],
            Array [
              Object {
                "size": Array [
                  "small",
                  false,
                  "large",
                  "huge",
                ],
              },
            ],
            Array [
              Object {
                "header": Array [
                  1,
                  2,
                  3,
                  4,
                  5,
                  6,
                  false,
                ],
              },
            ],
            Array [
              Object {
                "color": Array [],
              },
              Object {
                "background": Array [],
              },
            ],
            Array [
              Object {
                "font": Array [],
              },
            ],
            Array [
              Object {
                "align": Array [],
              },
            ],
            Array [
              "clean",
            ],
          ],
        }
      }
      name="content"
      onChange={[Function]}
      placeholder="Start typing here..."
      theme="snow"
      value=""
    />
  </div>
  <div
    className="css-55tkac"
  >
    <button
      className="css-1pcrr5b"
      type="submit"
    >
      Create
    </button>
  </div>
</form>
`;
